### ISSUE TITLE
Assertion with return doesn't reduce possible types

### ISSUE THREAD

**Issue by lptn** [vimeo/psalm#9200]
I'm trying to write stubs for Laravel code and got an issue that I  don't know how to solve. I tried different approaches, probably at least one of them should be valid and supported

`@return empty-mixed`:
https://psalm.dev/r/e00b8803da

plus `@psalm-assert empty-mixed`
https://psalm.dev/r/1f0ffd806d

plus template (does it work with in Intersection types like `@return TCondition&empty-mixed`?)
https://psalm.dev/r/655a4c75cb

**Comment by weirdan** [comment]
I think you need something like this: https://psalm.dev/r/e4610bf8f5

**Comment by lptn** [comment]
@weirdan 
Thanks. Then, in order to support all empty types, a construction like `@return (T is non-empty-mixed ? never : T)` should also work:

https://psalm.dev/r/7c664ac433

**Comment by weirdan** [comment]
You can invert the ternary and drop the negation in the conditional return type: https://psalm.dev/r/58f5da6efe

**Comment by lptn** [comment]
@weirdan 
G-R-E-A-T! Thanks a lot!

Does it mean `non-empty-mixed` is not fully supported psalm-type? Is this a bug?

Also, I created a stub for the opposite function `throw_unless`, it works, but I found 
```
/** @return positive-int|negative-int **/
```

https://psalm.dev/r/ca880864a1

doesn't work. Not sure is that a bug or not. Is it possible to easily specify int that not null? Like `non-falsy-int`?

**Comment by weirdan** [comment]
> Does it mean non-empty-mixed is not fully supported psalm-type? Is this a bug?

It seems there are quite a few types recognized everywhere but in conditional types.

>  `/** @return positive-int|negative-int **/` doesn't work.

The problem is two-fold:
1. Psalm has a limitation making it collapse integer ranges into a single range (which is not entirely correct). @orklah, the original author of int ranges, should know more about it.
2. There's a bug that causes `int<min,max>` not to be considered equivalent to `int`, as your example demonstrates. 

To sum up, currently, there's no type in the Psalm type system that could represent a set of integers except `0` across scope boundaries. `int<min,-1>|int<1,max>` could be that type, but it's not supported yet.

**Comment by orklah** [comment]
If you want to have joint ranges, this is the part that must be reworked: https://github.com/vimeo/psalm/blob/1341c15edcaf7146b9ed259d2b92b50d1f9018c7/src/Psalm/Internal/Type/TypeCombiner.php#L1159

This class is the TypeCombiner, it takes multiples types and outputs a single normalized Union (it's the thing that prevents ending up with types that looks like `lowercase-string|'someString'|string` because it collapse everything under `string`).

For now, the code stores literals on one side and tries to build some Int type with the non-literal types. The current goal was to build either a list of literals or a single type that would keep a maximum of information. Basically, a lot of things were collapsed under an integer range when possible but things like `8|int<80,90>` are currently collapsed to `int<8,90>`

**Comment by orklah** [comment]
Note: the `int<min, max>` into `int` is a bug in ScalarTypeComparator:
https://github.com/vimeo/psalm/blob/1c19260cddea29803cba2ef31c3a49189bcdeb0c/src/Psalm/Internal/Type/Comparator/ScalarTypeComparator.php#L364

The condition above was made exactly for that. It states that if we have an Int input and a IntRange as a container, the only way to not have a coercion is that the range has no bounds. This is correct, but there is no return true here to allow the case. Everything falls in the return false

Adding a `else { return true; }` solves the issue
